File: utils/benchmark.py
import torch
import torch.nn as nn
import time
import random
import numpy as np

# why cpu? 
# 1) i don't have a gpu 
# 2) go-torch only supports cpu 
DEVICE = torch.device("cpu")
print(f"PyTorch using device: {DEVICE}")




# bench params 
NUM_ITERATIONS = 100
SMALL_DIM = 128
MEDIUM_DIM = 512
LARGE_DIM = 1024
DEFAULT_BATCH_SIZE = 32




# generate a random pytorch tensor 
def generate_random_tensor_pytorch(shape, requires_grad=False):
    return torch.randn(shape, device=DEVICE, requires_grad=requires_grad)

def benchmark_tensor_creation_pytorch(dim, iterations):
    total_duration = 0
    shape = (dim, dim)
    for _ in range(iterations):
        start = time.perf_counter()
        _ = torch.randn(shape, device=DEVICE) 
        total_duration += (time.perf_counter() - start)
    return (total_duration / iterations) * 1000 





# we perform the following dimension-based tasks (for 100 iterations):
# 1) element wise addition        - 128, 512, 1024 
# 2) element wise multiplication  - 128, 512, 1024 
# 3) matrix multiplication        - 128, 512, 1024 
# 4) relu activation 

# all these metrics are calculated as ms (time * 1000). 

def benchmark_add_pytorch(dim, iterations):
    total_duration = 0
    tA = generate_random_tensor_pytorch((dim, dim))
    tB = generate_random_tensor_pytorch((dim, dim))
    for _ in range(iterations):
        start = time.perf_counter()
        _ = tA + tB
        total_duration += (time.perf_counter() - start)
    return (total_duration / iterations) * 1000 

def benchmark_mul_pytorch(dim, iterations):
    total_duration = 0
    tA = generate_random_tensor_pytorch((dim, dim))
    tB = generate_random_tensor_pytorch((dim, dim))
    for _ in range(iterations):
        start = time.perf_counter()
        _ = tA * tB
        total_duration += (time.perf_counter() - start)
    return (total_duration / iterations) * 1000 

def benchmark_matmul_pytorch(m, k, n, iterations):
    total_duration = 0
    tA = generate_random_tensor_pytorch((m, k))
    tB = generate_random_tensor_pytorch((k, n))
    for _ in range(iterations):
        start = time.perf_counter()
        _ = torch.matmul(tA, tB)
        total_duration += (time.perf_counter() - start)
    return (total_duration / iterations) * 1000 

def benchmark_relu_pytorch(dim, iterations):
    total_duration = 0
    t = generate_random_tensor_pytorch((dim, dim))
    for _ in range(iterations):
        start = time.perf_counter()
        _ = torch.relu(t)
        total_duration += (time.perf_counter() - start)
    return (total_duration / iterations) * 1000





# we perform the following layer and loss benchmarks 
# 1) linear layer forward   
# 2) cross-Entropy Loss     
# 3) Full-forward Backward  

def benchmark_linear_forward_pytorch(batch_size, input_dim, output_dim, iterations):
    total_duration = 0
    linear_layer = nn.Linear(input_dim, output_dim).to(DEVICE)
    input_tensor = generate_random_tensor_pytorch((batch_size, input_dim))
    for _ in range(iterations):
        start = time.perf_counter()
        _ = linear_layer(input_tensor)
        total_duration += (time.perf_counter() - start)
    return (total_duration / iterations) * 1000 

def benchmark_cross_entropy_loss_pytorch(batch_size, num_classes, iterations):
    total_duration = 0
    loss_fn = nn.CrossEntropyLoss()
    logits = generate_random_tensor_pytorch((batch_size, num_classes))
    targets = torch.randint(0, num_classes, (batch_size,), device=DEVICE)
    for _ in range(iterations):
        start = time.perf_counter()
        _ = loss_fn(logits, targets)
        total_duration += (time.perf_counter() - start)
    return (total_duration / iterations) * 1000 

def benchmark_forward_backward_pytorch(batch_size, input_dim, hidden_dim, output_dim, iterations):
    total_duration = 0
    
    for _ in range(iterations):
        x = generate_random_tensor_pytorch((batch_size, input_dim), requires_grad=True)
        layer1 = nn.Linear(input_dim, hidden_dim).to(DEVICE)
        layer2 = nn.Linear(hidden_dim, output_dim).to(DEVICE)
        loss_fn = nn.CrossEntropyLoss()
        targets = torch.randint(0, output_dim, (batch_size,), device=DEVICE)

        start = time.perf_counter()

        # forward pass
        h = layer1(x)
        h_relu = torch.relu(h)
        logits = layer2(h_relu)
        loss = loss_fn(logits, targets)

        # backward pass
        if x.grad is not None: x.grad.zero_()
        for param in layer1.parameters():
            if param.grad is not None: param.grad.zero_()
        for param in layer2.parameters():
            if param.grad is not None: param.grad.zero_()
        
        loss.backward()
        total_duration += (time.perf_counter() - start)
        # we end the timer here
        
    return (total_duration / iterations) * 1000 




# the print statements are self-explanatory. maybe try increasing the num of increasing, if you feel the test is unstable

if __name__ == "__main__":
    torch.manual_seed(int(time.time()))
    np.random.seed(int(time.time()))
    random.seed(int(time.time()))

    print("--- PyTorch Benchmarks (CPU) ---")
    print(f"Iterations per benchmark: {NUM_ITERATIONS}\n")

    dims_to_test = [SMALL_DIM, MEDIUM_DIM, LARGE_DIM]
    for d in dims_to_test:
        print(f"--- Dimension: {d}x{d} ---")
        print(f"Element-wise Add: {benchmark_add_pytorch(d, NUM_ITERATIONS):.3f} ms")
        print(f"Element-wise Mul: {benchmark_mul_pytorch(d, NUM_ITERATIONS):.3f} ms")
        print(f"Matrix Multiply ({d}x{d} @ {d}x{d}): {benchmark_matmul_pytorch(d, d, d, NUM_ITERATIONS):.3f} ms")
        print(f"ReLU Activation: {benchmark_relu_pytorch(d, NUM_ITERATIONS):.3f} ms")
        print()

    print("--- Layer and Loss Benchmarks ---")
    input_dim = 128
    hidden_dim = 256
    output_dim = 10  

    print(f"Linear Layer Forward (Batch: {DEFAULT_BATCH_SIZE}, In: {input_dim}, Out: {output_dim}): "
          f"{benchmark_linear_forward_pytorch(DEFAULT_BATCH_SIZE, input_dim, output_dim, NUM_ITERATIONS):.3f} ms")

    print(f"CrossEntropyLoss (Batch: {DEFAULT_BATCH_SIZE}, Classes: {output_dim}): "
          f"{benchmark_cross_entropy_loss_pytorch(DEFAULT_BATCH_SIZE, output_dim, NUM_ITERATIONS):.3f} ms")
    
    fwd_bwd_iterations = NUM_ITERATIONS // 10
    if fwd_bwd_iterations < 10: 
        fwd_bwd_iterations = 10 
        
    print(f"Full Forward-Backward (Net: {input_dim}-{hidden_dim}-{output_dim}, Batch: {DEFAULT_BATCH_SIZE}, Iter: {fwd_bwd_iterations}): "
          f"{benchmark_forward_backward_pytorch(DEFAULT_BATCH_SIZE, input_dim, hidden_dim, output_dim, fwd_bwd_iterations):.3f} ms")


    print("\n--- Benchmarks Complete ---")

File: utils/benchmark.go
package main

import (
	"fmt"
	"log"
	"math/rand"
	"time"

	"go-torch/nn"
	"go-torch/tensor"
)



// bench params 
const (
	numIterations    = 100 
	smallDim         = 128
	mediumDim        = 512
	largeDim         = 1024
	defaultBatchSize = 32
)


// utility functions for benchmark creation 

func generateRandomData(size int) []float64 {
	data := make([]float64, size)
	for i := range data {
		data[i] = rand.Float64()*2 - 1 
	}
	return data
}


// we perform the following dimension-based tasks (for 100 iterations):
// 1) element wise addition        - 128, 512, 1024 
// 2) element wise multiplication  - 128, 512, 1024 
// 3) matrix multiplication        - 128, 512, 1024 
// 4) relu activation 

// all these metrics are calculated as ms (time * 1000). 

func benchmarkAdd_GoTorch(dim int, iterations int) time.Duration {
	var totalDuration time.Duration
	shape := []int{dim, dim}
	numElements := dim * dim
	dataA := generateRandomData(numElements)
	dataB := generateRandomData(numElements)
	tA, _ := tensor.NewTensor(shape, dataA)
	tB, _ := tensor.NewTensor(shape, dataB)

	for i := 0; i < iterations; i++ {
		start := time.Now()
		_, err := tensor.AddTensor(tA, tB)
		if err != nil {
			log.Fatalf("GoTorch: Error in AddTensor: %v", err)
		}
		totalDuration += time.Since(start)
	}
	return totalDuration / time.Duration(iterations)
}

func benchmarkMul_GoTorch(dim int, iterations int) time.Duration {
	var totalDuration time.Duration
	shape := []int{dim, dim}
	numElements := dim * dim
	dataA := generateRandomData(numElements)
	dataB := generateRandomData(numElements)
	tA, _ := tensor.NewTensor(shape, dataA)
	tB, _ := tensor.NewTensor(shape, dataB)

	for i := 0; i < iterations; i++ {
		start := time.Now()
		_, err := tensor.MulTensor(tA, tB)
		if err != nil {
			log.Fatalf("GoTorch: Error in MulTensor: %v", err)
		}
		totalDuration += time.Since(start)
	}
	return totalDuration / time.Duration(iterations)
}

func benchmarkMatMul_GoTorch(m, k, n int, iterations int) time.Duration {
	var totalDuration time.Duration
	shapeA := []int{m, k}
	shapeB := []int{k, n}
	dataA := generateRandomData(m * k)
	dataB := generateRandomData(k * n)
	tA, _ := tensor.NewTensor(shapeA, dataA)
	tB, _ := tensor.NewTensor(shapeB, dataB)

	for i := 0; i < iterations; i++ {
		start := time.Now()
		_, err := tensor.MatMulTensor(tA, tB)
		if err != nil {
			log.Fatalf("GoTorch: Error in MatMulTensor: %v", err)
		}
		totalDuration += time.Since(start)
	}
	return totalDuration / time.Duration(iterations)
}

func benchmarkReLU_GoTorch(dim int, iterations int) time.Duration {
	var totalDuration time.Duration
	shape := []int{dim, dim}
	numElements := dim * dim
	data := generateRandomData(numElements)
	t, _ := tensor.NewTensor(shape, data)

	for i := 0; i < iterations; i++ {
		start := time.Now()
		_, err := nn.RELU(t)
		if err != nil {
			log.Fatalf("GoTorch: Error in RELU: %v", err)
		}
		totalDuration += time.Since(start)
	}
	return totalDuration / time.Duration(iterations)
}



// we perform the following layer and loss benchmarks 
// 1) linear layer forward   
// 2) cross-Entropy Loss     
// 3) Full-forward Backward  

func benchmarkLinearForward_GoTorch(batchSize, inputDim, outputDim int, iterations int) time.Duration {
	var totalDuration time.Duration
	linearLayer, err := nn.NewLinear(inputDim, outputDim)
	if err != nil {
		log.Fatalf("GoTorch: Error creating linear layer: %v", err)
	}
	inputData := generateRandomData(batchSize * inputDim)
	inputTensor, _ := tensor.NewTensor([]int{batchSize, inputDim}, inputData)

	for i := 0; i < iterations; i++ {
		start := time.Now()
		_, err := linearLayer.Forward(inputTensor)
		if err != nil {
			log.Fatalf("GoTorch: Error in Linear Forward: %v", err)
		}
		totalDuration += time.Since(start)
	}
	return totalDuration / time.Duration(iterations)
}

func benchmarkCrossEntropyLoss_GoTorch(batchSize, numClasses int, iterations int) time.Duration {
	var totalDuration time.Duration
	logitsData := generateRandomData(batchSize * numClasses)
	logits, _ := tensor.NewTensor([]int{batchSize, numClasses}, logitsData)
	targets := make([]int, batchSize)
	for i := range targets {
		targets[i] = rand.Intn(numClasses)
	}

	for i := 0; i < iterations; i++ {
		start := time.Now()
		_, err := nn.CrossEntropyLoss(logits, targets)
		if err != nil {
			log.Fatalf("GoTorch: Error in CrossEntropyLoss: %v", err)
		}
		totalDuration += time.Since(start)
	}
	return totalDuration / time.Duration(iterations)
}

func benchmarkForwardBackward_GoTorch(batchSize, inputDim, hiddenDim, outputDim int, iterations int) time.Duration {
	var totalDuration time.Duration

	for i := 0; i < iterations; i++ {
		inputData := generateRandomData(batchSize * inputDim)
		x, _ := tensor.NewTensor([]int{batchSize, inputDim}, inputData)
		x.RequiresGrad = true

		layer1, _ := nn.NewLinear(inputDim, hiddenDim)
		layer2, _ := nn.NewLinear(hiddenDim, outputDim)

		targets := make([]int, batchSize)
		for i := range targets {
			targets[i] = rand.Intn(outputDim)
		}

		// begin the timer here
		start := time.Now()

		// forward pass
		h, _ := layer1.Forward(x)
		hRelu, _ := nn.RELU(h)
		logits, _ := layer2.Forward(hRelu)
		loss, err := nn.CrossEntropyLoss(logits, targets)
		if err != nil {
			log.Fatalf("GoTorch: Error in loss during forward/backward: %v", err)
		}

		// backward pass
		if loss.RequiresGrad {
			loss.Backward(nil)
		}
		totalDuration += time.Since(start)
		// end there timer here

		// zero grads are optional 
		x.ZeroGrad()
		layer1.ZeroGrad()
		layer2.ZeroGrad()
	}
	return totalDuration / time.Duration(iterations)
}

func main() {
	rand.Seed(time.Now().UnixNano())  // deprecated but works, ig. you could change it. :)
	fmt.Println("--- Go-Torch Benchmarks ---")
	fmt.Printf("Iterations per benchmark: %d\n\n", numIterations)

	dims := []int{smallDim, mediumDim, largeDim}
	for _, d := range dims {
		fmt.Printf("--- Dimension: %dx%d ---\n", d, d)
		fmt.Printf("Element-wise Add: %v\n", benchmarkAdd_GoTorch(d, numIterations))
		fmt.Printf("Element-wise Mul: %v\n", benchmarkMul_GoTorch(d, numIterations))
		fmt.Printf("Matrix Multiply (%dx%d @ %dx%d): %v\n", d, d, d, d, benchmarkMatMul_GoTorch(d, d, d, numIterations))
		fmt.Printf("ReLU Activation: %v\n", benchmarkReLU_GoTorch(d, numIterations))
		fmt.Println()
	}

	fmt.Println("--- Layer and Loss Benchmarks ---")
	inputDim := 128
	hiddenDim := 256
	outputDim := 10 

	fmt.Printf("Linear Layer Forward (Batch: %d, In: %d, Out: %d): %v\n",
		defaultBatchSize, inputDim, outputDim,
		benchmarkLinearForward_GoTorch(defaultBatchSize, inputDim, outputDim, numIterations))

	fmt.Printf("CrossEntropyLoss (Batch: %d, Classes: %d): %v\n",
		defaultBatchSize, outputDim,
		benchmarkCrossEntropyLoss_GoTorch(defaultBatchSize, outputDim, numIterations))

	fmt.Printf("Full Forward-Backward (Net: %d-%d-%d, Batch: %d): %v\n",
		inputDim, hiddenDim, outputDim, defaultBatchSize,
		benchmarkForwardBackward_GoTorch(defaultBatchSize, inputDim, hiddenDim, outputDim, numIterations/10)) // Reduced iterations for slower Fwd-Bwd

	fmt.Println("\n--- Benchmarks Complete ---")
}